### PR TITLE
Further improvements to `go generate` for geth contract wrappers

### PR DESCRIPTION
**Pull request by coventry**
- LINK generation was exhibiting the same bash redirection issue as the general
  case, so that's in go now, too.

- There are intermittent cases where the EVM requires the metadata suffix on the
  contract bytecode, so a constant metadata suffix is appended to the bytecode
  for the contract wrappers. This avoids unnecessary git diffs.

### CODE REVIEW

File: core/internal/gethwrappers/abigen.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+// arg.
+type AbigenArgs struct {
+	Bin, ABI, Out, Type, Pkg string
+}
+
+// Abigen calls Abigen  with the given arguments
+//
+// It might seem like a shame, to shell out to another golang program like
+// this, but the abigen executable is the stable public interface to the
+// geth contract-wrapper machinery.
```
Comment (samsondav) [inline core/internal/gethwrappers/abigen.go:23]: Wild

File: core/internal/gethwrappers/extract_contract_details.go [hunk 2]
```diff
@@ -84,3 +85,9 @@ func ExtractContractDetails(beltArtifactPath string) (*ContractDetails, error) {
 var binarySuffixRegexp = regexp.MustCompile(
 	"^a264697066735822[[:xdigit:]]{68}64736f6c6343[[:xdigit:]]{6}0033$",
 ).MatchString
+
+// constantBinaryMetadataSuffix is a constant stand-in for the metadata suffix
+// which the EVM expects (and in some cases, it seems, requires) to find at the
+// end of the binary object representing a contract under deployment.
+var constantBinaryMetadataSuffix = "a264697066735822" +
```
Comment (samsondav) [inline core/internal/gethwrappers/extract_contract_details.go:92]: Bizarre... where does this come from out of interest?

File: core/internal/gethwrappers/generation/generate_link/wrap_link.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	fmt.Println("Generating", pkgName, "contract wrapper")
+	className := "LinkToken"
+	tmpDir, cleanup := gethwrappers.TempDir(className)
+	defer cleanup()
+	linkDetails, err := ioutil.ReadFile(filepath.Join(
+		gethwrappers.GetProjectRoot(), "evm-test-helpers/src/LinkToken.json"))
+	if err != nil {
+		gethwrappers.Exit("could not read LINK contract details", err)
+	}
+	if fmt.Sprintf("%x", sha256.Sum256(linkDetails)) !=
```
Comment (samsondav) [inline core/internal/gethwrappers/generation/generate_link/wrap_link.go:39]: `%x` is for hex? I haven't seen that before

File: core/internal/gethwrappers/utils.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	}
+	for root != "/" { // Walk up path to find dir containing go.mod
+		if _, err := os.Stat(filepath.Join(root, "go.mod")); os.IsNotExist(err) {
+			root = filepath.Dir(root)
+		} else {
+			return root
+		}
+	}
+	Exit("could not find project root", nil)
+	panic("can't get here")
```
Comment (samsondav) [inline core/internal/gethwrappers/utils.go:34]: What is this for?

Comment (coventry) [reply core/internal/gethwrappers/utils.go:34]: It satisfies staticcheck, which complains that there's no return at the end of the function.

File: core/internal/gethwrappers/extract_contract_details.go (hunk 2 above)
Comment (coventry) [reply core/internal/gethwrappers/extract_contract_details.go:92]: I should have included a link to this in the comment. I'll do it in a later PR. https://solidity.readthedocs.io/en/v0.5.3/metadata.html

File: core/internal/gethwrappers/abigen.go (hunk 1 above)
Comment (coventry) [reply core/internal/gethwrappers/abigen.go:23]: To be clear, there are "public" interfaces to the underlying golang functions, in the sense that  they're not in `internal` packages and start with capital letters. I just expect that `abigen`'s CLI interface will be more stable than those.

File: core/internal/gethwrappers/generation/generate_link/wrap_link.go (hunk 3 above)
Comment (coventry) [reply core/internal/gethwrappers/generation/generate_link/wrap_link.go:39]: Yeah, it's nice.

**Pull request merged**